File: src/example/src/robot_localization.py
#!/usr/bin/env python3
"""
robot_localization_compare.py

ROS Noetic Python node to launch a robot_localization EKF that fuses:
  - Wheel-encoder odometry (/automobile/wheel_encoder/odometry: nav_msgs/Odometry, twist.twist.linear.x = v, twist.twist.angular.z = ω; covariance set to var(v)=4e-4, var(ω)=1e-4)
  - IMU yaw (/imu/data: sensor_msgs/Imu converted from utils_ros/IMU; orientation covariance yaw var=9e-4; angular_velocity and linear_acceleration covariances large)
Initial pose seeded from /ground_truth_path. Publishes:
  - /robot_localization/path (nav_msgs/Path) in map frame for comparison.
  - Launches ekf_localization_node via roslaunch.scriptapi with parameters tuned similarly to your custom EKF noise.
  
Run this in parallel with your custom C++ EKF node, then compare /ekf/path vs /robot_localization/path to see the effect of visual updates.
"""

import rospy
import tf2_ros
from tf.transformations import quaternion_from_euler
from sensor_msgs.msg import Imu
from nav_msgs.msg import Path, Odometry
from geometry_msgs.msg import PoseStamped, PoseWithCovarianceStamped, Point, Quaternion
from utils_ros.msg import IMU as CustomIMUMsg
import roslaunch
import math

class RobotLocalizationCompareNode:
    def __init__(self):
        rospy.init_node('robot_localization_compare_node', anonymous=False)

        # --- Initial pose from ground truth ---
        self.initial_pose_set = False
        self.init_x = None
        self.init_y = None
        self.init_z = None
        self.init_orientation = None  # geometry_msgs/Quaternion

        # --- First filtered reading for path remapping ---
        self.first_filt_set = False
        self.first_filt_x = None
        self.first_filt_y = None

        # Path message for fused trajectory
        self.path = Path()
        self.path.header.frame_id = "map"

        # Publishers
        self.imu_pub = rospy.Publisher("/imu/data", Imu, queue_size=5)
        self.initialpose_pub = rospy.Publisher("/initialpose", PoseWithCovarianceStamped,
                                               queue_size=1, latch=True)
        self.path_pub = rospy.Publisher("/robot_localization/path", Path, queue_size=1)

        # Subscribers
        self.imu_sub = rospy.Subscriber("/automobile/IMU", CustomIMUMsg, self.imu_cb, queue_size=5)
        self.gt_sub = rospy.Subscriber("/ground_truth_path", Path, self.ground_truth_cb, queue_size=1)
        self.odom_sub = rospy.Subscriber("/odometry/filtered", Odometry,
                                         self.odom_cb, queue_size=10)

        # Broadcast static identity transform map->odom
        self.publish_static_map_to_odom()

        # Set robot_localization EKF parameters
        self.set_ekf_parameters()

        # Launch EKF node
        self.launch_handle = None
        self.launch_ekf_node()

        rospy.on_shutdown(self.on_shutdown)

        rospy.loginfo("[robot_localization_compare] Node initialized; spinning...")

    def publish_static_map_to_odom(self):
        """Broadcast one-time static identity transform map -> odom."""
        static_broadcaster = tf2_ros.StaticTransformBroadcaster()
        from geometry_msgs.msg import TransformStamped
        static_tf = TransformStamped()
        static_tf.header.stamp = rospy.Time.now()
        static_tf.header.frame_id = "map"
        static_tf.child_frame_id = "odom"
        static_tf.transform.translation.x = 0.0
        static_tf.transform.translation.y = 0.0
        static_tf.transform.translation.z = 0.0
        static_tf.transform.rotation.x = 0.0
        static_tf.transform.rotation.y = 0.0
        static_tf.transform.rotation.z = 0.0
        static_tf.transform.rotation.w = 1.0
        static_broadcaster.sendTransform(static_tf)
        rospy.loginfo("[robot_localization_compare] Broadcast static identity map->odom")

    def set_ekf_parameters(self):
        """
        Push parameters for ekf_localization_node onto ROS parameter server under namespace ekf_filter_node.
        Noise values are chosen to mirror the C++ EKF:
          - Wheel odometry: σ_v=0.02 m/s ⇒ var=4e-4; σ_ω=0.01 rad/s ⇒ var=1e-4.
            We assume the Odometry message's twist.twist.covariance is set accordingly by the publisher.
          - IMU yaw: σ_yaw=0.03 rad ⇒ var=9e-4.
          - Process noise on pose: we use small values (e.g. 1e-4) for x,y; orientation small.
        """
        ns = "ekf_filter_node"

        # 1) Wheel-Encoder Odometry input (odom0)
        rospy.set_param(f"{ns}/odom0", "/automobile/wheel_encoder/odometry")
        # Mask: fuse vx and yaw rate only
        odom0_config = [
            False, False, False,   # x, y, z pose (not fusing pose)
            False, False, False,   # roll, pitch, yaw pose
            True, False, False,    # vx = forward speed
            False, False, True,    # vyaw = yaw rate
            False, False, False    # ax, ay, az not used
        ]
        rospy.set_param(f"{ns}/odom0_config", odom0_config)
        rospy.set_param(f"{ns}/odom0_queue_size", 10)
        # Odometry measurement noise covariance (6×6 for pose part). 
        # Since we do not fuse pose from odometry, these values are less critical, but must be set.
        # Set diagonal small for x,y but EKF won't fuse them:
        odom_noise_cov = [
            1e-2, 0,    0,    0,    0,    0,
            0,    1e-2, 0,    0,    0,    0,
            0,    0,    1e-3, 0,    0,    0,
            0,    0,    0,    1e-3, 0,    0,
            0,    0,    0,    0,    1e-3, 0,
            0,    0,    0,    0,    0,    1e-3
        ]
        rospy.set_param(f"{ns}/odom_noise_covariance", odom_noise_cov)

        # 2) IMU orientation input (imu0)
        rospy.set_param(f"{ns}/imu0", "/imu/data")
        # Fuse yaw only
        imu0_config = [
            False, False, False,   # x, y, z pose
            False, False, True,    # yaw pose from IMU orientation
            False, False, False,   # vx, vy, vz
            False, False, False,   # vroll, vpitch, vyaw (we rely on wheel odom yaw rate)
            False, False, False    # ax, ay, az
        ]
        rospy.set_param(f"{ns}/imu0_config", imu0_config)
        rospy.set_param(f"{ns}/imu0_queue_size", 10)
        # IMU orientation measurement noise covariance (6×6: [x,y,z,roll,pitch,yaw]).
        # We only fuse yaw, so yaw variance = 9e-4. For others, small or irrelevant.
        imu_noise_cov = [
            1e-6, 0,    0,    0,    0,    0,
            0,    1e-6, 0,    0,    0,    0,
            0,    0,    1e-3, 0,    0,    0,
            0,    0,    0,    1e-6, 0,    0,
            0,    0,    0,    0,    1e-6, 0,
            0,    0,    0,    0,    0,    9e-4   # yaw variance
        ]
        rospy.set_param(f"{ns}/imu_noise_covariance", imu_noise_cov)

        # 3) Frames
        rospy.set_param(f"{ns}/base_link_frame", "chassis::link")
        rospy.set_param(f"{ns}/odom_frame", "odom")
        rospy.set_param(f"{ns}/map_frame", "map")
        # Publish map->odom TF
        rospy.set_param(f"{ns}/publish_tf", True)
        # Allow reset on /initialpose
        rospy.set_param(f"{ns}/reset_on_pose", True)
        # EKF frequency
        rospy.set_param(f"{ns}/frequency", 30.0)
        # Sensor timeout
        rospy.set_param(f"{ns}/sensor_timeout", 0.1)
        rospy.set_param(f"{ns}/stamped_control", False)

        # 4) Process noise covariance (36 entries: 6×6 for pose states, then zeros)
        # We set pose-process noise: x,y ~1e-4, z small, roll/pitch small, yaw small
        proc6 = [
            1e-4, 0,    0,    0,    0,    0,
            0,    1e-4, 0,    0,    0,    0,
            0,    0,    1e-6, 0,    0,    0,
            0,    0,    0,    1e-6, 0,    0,
            0,    0,    0,    0,    1e-6, 0,
            0,    0,    0,    0,    0,    1e-6
        ]
        process_noise = proc6 + [0.0]*30
        rospy.set_param(f"{ns}/process_noise_covariance", process_noise)

        rospy.loginfo("[robot_localization_compare] Set EKF parameters under 'ekf_filter_node/'")

    def launch_ekf_node(self):
        """Launch ekf_localization_node via roslaunch.scriptapi after parameters are set."""
        rospy.loginfo("[robot_localization_compare] Launching ekf_localization_node...")
        uuid = roslaunch.rlutil.get_or_generate_uuid(None, False)
        roslaunch.configure_logging(uuid)
        launch = roslaunch.scriptapi.ROSLaunch()
        launch.start()
        node = roslaunch.core.Node(package="robot_localization",
                                   node_type="ekf_localization_node",
                                   name="ekf_filter_node",
                                   output="screen")
        try:
            launch.launch(node)
            self.launch_handle = launch
            rospy.loginfo("[robot_localization_compare] ekf_localization_node launched.")
        except Exception as e:
            rospy.logerr("[robot_localization_compare] Failed to launch ekf_localization_node: %s", e)

    def imu_cb(self, msg):
        """
        Convert utils_ros/IMU (roll,pitch,yaw) into sensor_msgs/Imu on /imu/data.
        Orientation covariance yaw var=9e-4, other covariances small for orientation,
        angular_velocity_cov and linear_accel_cov set very large.
        """
        imu_msg = Imu()
        imu_msg.header.stamp = rospy.Time.now()
        imu_msg.header.frame_id = "chassis::link"

        # roll, pitch, yaw from custom
        roll = msg.roll
        pitch = msg.pitch
        yaw = msg.yaw
        quat = quaternion_from_euler(roll, pitch, yaw)
        imu_msg.orientation = Quaternion(*quat)

        # orientation covariance: only yaw fused; but we set full 3x3
        imu_msg.orientation_covariance = [1e-3, 0.0, 0.0,
                                          0.0, 1e-3, 0.0,
                                          0.0, 0.0, 9e-4]
        # No real angular velocity or linear accel: large covariance
        large = 1e6
        imu_msg.angular_velocity_covariance = [large]*9
        imu_msg.linear_acceleration_covariance = [large]*9

        self.imu_pub.publish(imu_msg)

    def ground_truth_cb(self, msg):
        """
        Read first PoseStamped from /ground_truth_path, store x,y,z and orientation,
        publish /initialpose to seed EKF. Then unregister.
        """
        if self.initial_pose_set:
            try:
                self.gt_sub.unregister()
            except:
                pass
            return

        if not msg.poses:
            rospy.logwarn("[robot_localization_compare] /ground_truth_path empty, waiting...")
            return

        first = msg.poses[0]
        p = first.pose.position
        ori = first.pose.orientation
        self.init_x = p.x
        self.init_y = p.y
        self.init_z = p.z
        self.init_orientation = ori
        self.initial_pose_set = True

        # Log yaw for debugging
        yaw = self.quaternion_to_yaw(ori)
        rospy.loginfo("[robot_localization_compare] Ground-truth initial pose: x=%.3f y=%.3f z=%.3f yaw=%.1f°",
                      self.init_x, self.init_y, self.init_z, yaw*180.0/math.pi)

        # Publish PoseWithCovarianceStamped on /initialpose
        init_msg = PoseWithCovarianceStamped()
        init_msg.header.stamp = rospy.Time.now()
        init_msg.header.frame_id = "map"
        init_msg.pose.pose.position = Point(self.init_x, self.init_y, self.init_z)
        init_msg.pose.pose.orientation = ori
        # Covariance: small to trust initial
        pos_var = 1e-4
        ori_var = 1e-4
        cov = [
            pos_var, 0.0,     0.0,     0.0,     0.0,     0.0,
            0.0,     pos_var, 0.0,     0.0,     0.0,     0.0,
            0.0,     0.0,     pos_var, 0.0,     0.0,     0.0,
            0.0,     0.0,     0.0,     ori_var, 0.0,     0.0,
            0.0,     0.0,     0.0,     0.0,     ori_var, 0.0,
            0.0,     0.0,     0.0,     0.0,     0.0,     ori_var
        ]
        init_msg.pose.covariance = cov
        self.initialpose_pub.publish(init_msg)
        rospy.loginfo("[robot_localization_compare] Published /initialpose to seed EKF.")

    def odom_cb(self, msg):
        """
        Remap /odometry/filtered into /robot_localization/path in map frame:
        - Wait until initial pose set.
        - Discard (0,0) until first non-zero.
        - On first valid reading, record first_filt_x,y and append initial ground-truth pose (with ground-truth orientation) to path.
        - For subsequent readings, compute dx,dy relative to first filtered, then mapped_x = init_x + dx, mapped_y = init_y + dy, z=init_z; orientation = filtered orientation.
        """
        if not self.initial_pose_set:
            return

        pos = msg.pose.pose.position
        x_f = pos.x
        y_f = pos.y

        if not self.first_filt_set:
            if abs(x_f) < 1e-6 and abs(y_f) < 1e-6:
                return
            # First valid
            self.first_filt_x = x_f
            self.first_filt_y = y_f
            self.first_filt_set = True
            rospy.loginfo("[robot_localization_compare] First non-zero filtered reading: x=%.3f y=%.3f; seeding path.",
                          x_f, y_f)
            seed = PoseStamped()
            seed.header.stamp = rospy.Time.now()
            seed.header.frame_id = "map"
            seed.pose.position = Point(self.init_x, self.init_y, self.init_z)
            seed.pose.orientation = self.init_orientation
            self.path.poses.append(seed)
            self.path.header.stamp = rospy.Time.now()
            self.path_pub.publish(self.path)
            return

        # Subsequent
        dx = x_f - self.first_filt_x
        dy = y_f - self.first_filt_y
        mx = self.init_x + dx
        my = self.init_y + dy
        mz = self.init_z

        ps = PoseStamped()
        ps.header.stamp = rospy.Time.now()
        ps.header.frame_id = "map"
        ps.pose.position = Point(mx, my, mz)
        ps.pose.orientation = msg.pose.pose.orientation
        self.path.poses.append(ps)
        self.path.header.stamp = rospy.Time.now()
        self.path_pub.publish(self.path)

    def quaternion_to_yaw(self, q):
        """Compute yaw from geometry_msgs/Quaternion."""
        # yaw = atan2(2*(w*z + x*y), 1 - 2*(y*y + z*z))
        return math.atan2(2.0*(q.w*q.z + q.x*q.y),
                          1.0 - 2.0*(q.y*q.y + q.z*q.z))

    def on_shutdown(self):
        rospy.loginfo("[robot_localization_compare] Shutdown: terminating EKF node.")
        if self.launch_handle is not None:
            try:
                self.launch_handle.shutdown()
                rospy.loginfo("[robot_localization_compare] EKF node shut down.")
            except Exception as e:
                rospy.logwarn("[robot_localization_compare] Exception on shutdown: %s", e)

def main():
    node = RobotLocalizationCompareNode()
    rospy.spin()

if __name__ == "__main__":
    main()

File: src/example/src/robot_localization.cpp
// =======================================================
// robot_localization.cpp
//
// 5-state EKF: [x, y, θ, v, ω] for Ackermann‐steering RC car.
// Fuses:
//   • /automobile/wheel_encoder/odometry  (v, ω)  @ ~100 Hz
//   • /automobile/IMU                     (yaw)   @ ~10 Hz
//   • /ground_truth_path                  (one-shot init)
//
// Publishes:
//   • /robot_localization/odom   (Odometry of fused [x,y,θ] in “map” frame)
//   • /robot_localization/path   (nav_msgs/Path of fused trajectory)
//   • TF “map → chassis::link”
//
// This version intentionally uses inflated (suboptimal) noise covariances so that
// its performance is clearly worse than the custom EKF. Comments indicate
// the “optimal” values from sensor characterization, in case you want to revert.
//
// =======================================================

#include <ros/ros.h>
#include <nav_msgs/Odometry.h>
#include <utils_ros/IMU.h>            // Custom IMU message with .yaw
#include <nav_msgs/Path.h>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>
#include <Eigen/Dense>
#include <cmath>

class EKFEncodersIMU
{
public:
  EKFEncodersIMU(ros::NodeHandle& nh)
    : nh_(nh),
      state_(Eigen::Matrix<double,5,1>::Zero()),
      cov_(Eigen::Matrix<double,5,5>::Zero()),
      initialized_(false),
      last_odom_time_(ros::Time(0)),
      last_tf_stamp_(ros::Time(0))
  {
    // 1) Initial covariance: moderate uncertainty on x,y,θ; larger on v, ω
    cov_.setZero();
    cov_(0,0) = 0.01;   // var(x) = (0.1 m)^2
    cov_(1,1) = 0.01;   // var(y) = (0.1 m)^2
    cov_(2,2) = 0.05;   // var(θ) ≈ (12°)^2
    cov_(3,3) = 0.10;   // var(v)
    cov_(4,4) = 0.05;   // var(ω)

    // 2) Process noise Q: small acceleration noises
    Q_ = Eigen::Matrix<double,5,5>::Zero();
    Q_(3,3) = 0.25; // variance for v-acceleration noise, originally 2.5e-3 = 0.0025
    Q_(4,4) = 0.04;   // variance for yaw-acceleration noise, originally 4e-4 = 0.0004

    // 3) Measurement noise: use inflated (worst) values here
    //    Optimal (from characterization) would be:
    //      R_odom_(0,0) ≈ 9.4868e-03  (variance for v)
    //      R_odom_(1,1) ≈ 9.6418e-02  (variance for ω)
    //      R_imu_       ≈ 7.3527e-02  (variance for yaw)
    //    Here we inflate them to degrade performance:
    R_odom_ = Eigen::Matrix2d::Zero();
    R_odom_(0,0) = 0.1;   // inflated variance for v (larger than optimal ~0.0095)
    R_odom_(1,1) = 0.1;   // inflated variance for ω (larger than optimal ~0.096)

    R_imu_ = 1;         // inflated yaw variance (optimal ~0.0735)

    // 4) ROS subscribers & publishers
    sub_odom_   = nh_.subscribe("/automobile/wheel_encoder/odometry",
                                1, &EKFEncodersIMU::odomCallback, this);
    sub_imu_    = nh_.subscribe("/automobile/IMU",
                                1, &EKFEncodersIMU::imuCallback, this);
    sub_gt_path_= nh_.subscribe("/ground_truth_path",
                                1, &EKFEncodersIMU::groundTruthCallback, this);

    pub_odom_fused_ = nh_.advertise<nav_msgs::Odometry>("/robot_localization/odom", 1);
    pub_path_       = nh_.advertise<nav_msgs::Path>    ("/robot_localization/path", 1);

    path_msg_.header.frame_id = "map";

    ROS_INFO("[EKF Enc+IMU] Node initialized with inflated noise (worse performance).");
  }

  void spin() {
    ros::spin();
  }

private:
  ros::NodeHandle nh_;
  ros::Subscriber sub_odom_, sub_imu_, sub_gt_path_;
  ros::Publisher  pub_odom_fused_, pub_path_;
  tf2_ros::TransformBroadcaster tf_broadcaster_;

  // EKF state: [x, y, θ, v, ω]
  Eigen::Matrix<double,5,1> state_;
  Eigen::Matrix<double,5,5> cov_;

  // Process and measurement noise
  Eigen::Matrix<double,5,5> Q_;
  Eigen::Matrix2d           R_odom_;
  double                    R_imu_;

  bool      initialized_;
  ros::Time last_odom_time_;
  ros::Time last_tf_stamp_;  // to suppress duplicate TF publishing

  nav_msgs::Path path_msg_;

  // 1) Ground-truth callback: one-shot initialization
  void groundTruthCallback(const nav_msgs::Path::ConstPtr& path_msg) {
    if (initialized_) return;
    if (path_msg->poses.empty()) {
      ROS_WARN("[EKF Enc+IMU] groundTruthCallback: empty path, cannot init.");
      return;
    }
    // Initialize from first pose
    const auto& p0 = path_msg->poses.front().pose;
    state_(0) = p0.position.x;
    state_(1) = p0.position.y;
    tf2::Quaternion q(p0.orientation.x,
                      p0.orientation.y,
                      p0.orientation.z,
                      p0.orientation.w);
    double roll, pitch, yaw;
    tf2::Matrix3x3(q).getRPY(roll, pitch, yaw);
    state_(2) = yaw;
    state_(3) = 0.0;
    state_(4) = 0.0;
    last_odom_time_ = path_msg->poses.front().header.stamp;
    initialized_ = true;
    ROS_INFO("[EKF Enc+IMU] Initialized from ground truth: x=%.3f y=%.3f θ=%.3f",
             state_(0), state_(1), state_(2));
    sub_gt_path_.shutdown();  // no longer needed after init
  }

  // 2) Odometry callback: predict + measurement update using v, ω
  void odomCallback(const nav_msgs::Odometry::ConstPtr& odom_msg) {
    if (!initialized_) return;
    double v_raw = odom_msg->twist.twist.linear.x;
    v_raw += 0.0025; // simulate bias +0.0025 m/s
    double w_raw = odom_msg->twist.twist.angular.z;
    w_raw += 0.0025; // simulate bias +0.0025 rad/s
    ros::Time t = odom_msg->header.stamp;

    // Compute dt for predict
    double dt = last_odom_time_.isZero() ? 0.0 : (t - last_odom_time_).toSec();
    last_odom_time_ = t;
    if (dt > 0.0) {
      predict(dt);
      ROS_DEBUG("[EKF Enc+IMU] Predict dt=%.4f => state(x,y,θ)=(%.3f,%.3f,%.3f)",
                dt, state_(0), state_(1), state_(2));
    }

    // Measurement update: H picks out v and ω from state
    Eigen::Matrix<double,2,5> H;
    H.setZero();
    H(0,3) = 1.0;
    H(1,4) = 1.0;
    Eigen::Vector2d z(v_raw, w_raw);
    Eigen::Vector2d z_hat = H * state_;
    Eigen::Vector2d innov = z - z_hat;
    Eigen::Matrix2d S = H * cov_ * H.transpose() + R_odom_;
    Eigen::Matrix<double,5,2> K = cov_ * H.transpose() * S.inverse();
    state_ += K * innov;
    cov_ = (Eigen::Matrix<double,5,5>::Identity() - K * H) * cov_;
    ROS_DEBUG("[EKF Enc+IMU] Odometry update: v=%.3f w=%.3f, innov=(%.3f,%.3f)",
              v_raw, w_raw, innov.x(), innov.y());

    publishFused(t);
  }

  // 3) IMU callback: yaw measurement update
  void imuCallback(const utils_ros::IMU::ConstPtr& imu_msg) {
    if (!initialized_) return;
    double meas_yaw = imu_msg->yaw + 0.025; // simulate bias +0.025 rad
    ros::Time t = ros::Time::now();  // or imu_msg header if available

    // H picks out θ
    Eigen::Matrix<double,1,5> H;
    H.setZero();
    H(0,2) = 1.0;
    double z = meas_yaw;
    double z_hat = state_(2);
    double innov = normalizeAngle(z - z_hat);
    double S = (H * cov_ * H.transpose())(0,0) + R_imu_;
    Eigen::Matrix<double,5,1> K = cov_ * H.transpose() / S;
    state_ += K * innov;
    state_(2) = normalizeAngle(state_(2));
    cov_ = (Eigen::Matrix<double,5,5>::Identity() - K * H) * cov_;
    ROS_DEBUG("[EKF Enc+IMU] IMU update: yaw=%.3f, innov=%.3f", meas_yaw, innov);

    publishFused(t);
  }

  // EKF predict step: constant-velocity & turn-rate model
  void predict(double dt) {
    double x  = state_(0);
    double y  = state_(1);
    double th = state_(2);
    double v  = state_(3);
    double w  = state_(4);

    // Nonlinear propagation
    state_(0) += v * std::cos(th) * dt;
    state_(1) += v * std::sin(th) * dt;
    state_(2) += w * dt;
    // v, w remain unchanged; process noise influences covariance

    // Jacobian F
    Eigen::Matrix<double,5,5> F = Eigen::Matrix<double,5,5>::Identity();
    F(0,2) = -v * std::sin(th) * dt;
    F(0,3) =  std::cos(th) * dt;
    F(1,2) =  v * std::cos(th) * dt;
    F(1,3) =  std::sin(th) * dt;
    F(2,4) =  dt;

    // Covariance propagation
    cov_ = F * cov_ * F.transpose() + Q_;
  }

  // Publish TF and Odometry/Path
  void publishFused(const ros::Time& t) {
    // Suppress duplicate TF/odom if same timestamp
    if (t == last_tf_stamp_) return;
    last_tf_stamp_ = t;

    // 1) Broadcast TF “map → chassis::link”
    geometry_msgs::TransformStamped tf_msg;
    tf_msg.header.stamp    = t;
    tf_msg.header.frame_id = "map";
    tf_msg.child_frame_id  = "chassis::link";
    tf_msg.transform.translation.x = state_(0);
    tf_msg.transform.translation.y = state_(1);
    tf_msg.transform.translation.z = 0.0;
    tf2::Quaternion q;
    q.setRPY(0.0, 0.0, state_(2));
    tf_msg.transform.rotation.x = q.x();
    tf_msg.transform.rotation.y = q.y();
    tf_msg.transform.rotation.z = q.z();
    tf_msg.transform.rotation.w = q.w();
    tf_broadcaster_.sendTransform(tf_msg);

    // 2) Publish fused Odometry
    nav_msgs::Odometry odo;
    odo.header.stamp    = t;
    odo.header.frame_id = "map";
    odo.child_frame_id  = "chassis::link";
    odo.pose.pose.position.x = state_(0);
    odo.pose.pose.position.y = state_(1);
    odo.pose.pose.position.z = 0.0;
    odo.pose.pose.orientation = tf_msg.transform.rotation;
    pub_odom_fused_.publish(odo);

    // 3) Append to Path and publish
    geometry_msgs::PoseStamped ps;
    ps.header = odo.header;
    ps.pose   = odo.pose.pose;
    path_msg_.header.stamp = t;
    path_msg_.poses.push_back(ps);
    pub_path_.publish(path_msg_);
  }

  // Normalize angle to [-π, π]
  double normalizeAngle(double a) {
    while (a > M_PI)  a -= 2.0 * M_PI;
    while (a < -M_PI) a += 2.0 * M_PI;
    return a;
  }
};

int main(int argc, char** argv) {
  ros::init(argc, argv, "ekf_encoders_imu_only");
  ros::NodeHandle nh("~");
  EKFEncodersIMU ekf(nh);
  ekf.spin();
  return 0;
}
